File: app/src/main/java/de/bahnhoefe/deutschlands/bahnhofsfotos/IntroSliderActivity.java
package de.bahnhoefe.deutschlands.bahnhofsfotos;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import de.bahnhoefe.deutschlands.bahnhofsfotos.databinding.ActivityIntroSliderBinding;

public class IntroSliderActivity extends AppCompatActivity {

    private ActivityIntroSliderBinding binding;
    private int[] layouts;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final BaseApplication baseApplication = (BaseApplication) getApplication();

        getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);

        binding = ActivityIntroSliderBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        layouts = new int[]{R.layout.intro_slider1, R.layout.intro_slider2};

        addBottomDots(0);
        changeStatusBarColor();
        final var viewPagerAdapter = new ViewPagerAdapter();
        binding.viewPager.setAdapter(viewPagerAdapter);
        binding.viewPager.addOnPageChangeListener(viewListener);

        binding.btnSliderSkip.setOnClickListener(v -> {
            baseApplication.setFirstAppStart(true);
            openMainActivity();
        });

        binding.btnSliderNext.setOnClickListener(v -> {
            final int current = getNextItem();
            if (current < layouts.length) {
                binding.viewPager.setCurrentItem(current);
            } else {
                openMainActivity();
            }
        });
    }

    private void openMainActivity() {
        final var intent = new Intent(IntroSliderActivity.this, MainActivity.class);
        startActivity(intent);
        finish();
    }

    @Override
    public void onBackPressed() {
        openMainActivity();
    }

    private void addBottomDots(final int position) {
        final var dots = new TextView[layouts.length];
        final var colorActive = getResources().getIntArray(R.array.dot_active);
        final var colorInactive = getResources().getIntArray(R.array.dot_inactive);
        binding.layoutDots.removeAllViews();

        for (int i = 0; i < dots.length; i++) {
            dots[i] = new TextView(this);
            dots[i].setText(Html.fromHtml("&#8226;", Html.FROM_HTML_MODE_LEGACY));
            dots[i].setTextSize(35);
            dots[i].setTextColor(colorInactive[position]);
            binding.layoutDots.addView(dots[i]);
        }

        if (dots.length > 0) {
            dots[position].setTextColor(colorActive[position]);
        }
    }

    private int getNextItem() {
        return binding.viewPager.getCurrentItem() + 1;
    }

    final ViewPager.OnPageChangeListener viewListener = new ViewPager.OnPageChangeListener() {

        @Override
        public void onPageScrolled(final int position, final float positionOffset, final int positionOffsetPixels) {

        }

        @Override
        public void onPageSelected(final int position) {
            final var baseApplication = (BaseApplication) getApplication();
            addBottomDots(position);

            if (position == layouts.length - 1) {
                binding.btnSliderNext.setText(R.string.proceed);
                binding.btnSliderSkip.setVisibility(View.INVISIBLE);
                baseApplication.setFirstAppStart(true);
            } else {
                binding.btnSliderNext.setText(R.string.next);
                binding.btnSliderSkip.setVisibility(View.VISIBLE);
            }
        }

        @Override
        public void onPageScrollStateChanged(final int state) {

        }
    };

    private void changeStatusBarColor() {
        final var window = getWindow();
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        window.setStatusBarColor(Color.TRANSPARENT);
    }

    public class ViewPagerAdapter extends PagerAdapter {

        @Override
        @NonNull
        public Object instantiateItem(@NonNull final ViewGroup container, final int position) {
            final var layoutInflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            final var view = layoutInflater.inflate(layouts[position], container, false);
            container.addView(view);
            return view;
        }

        @Override
        public void destroyItem(final ViewGroup container, final int position, @NonNull final Object object) {
            final var view = (View) object;
            container.removeView(view);
        }

        @Override
        public int getCount() {
            return layouts.length;
        }

        @Override
        public boolean isViewFromObject(@NonNull final View view, @NonNull final Object object) {
            return view == object;
        }
    }

}

File: app/src/main/kotlin/de/bahnhoefe/deutschlands/bahnhofsfotos/IntroSliderActivity.kt
package de.bahnhoefe.deutschlands.bahnhofsfotos

import android.content.Intent
import android.os.Bundle
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.MarginLayoutParams
import android.widget.TextView
import androidx.activity.OnBackPressedCallback
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updateLayoutParams
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import dagger.hilt.android.AndroidEntryPoint
import de.bahnhoefe.deutschlands.bahnhofsfotos.databinding.ActivityIntroSliderBinding
import de.bahnhoefe.deutschlands.bahnhofsfotos.util.PreferencesService
import javax.inject.Inject


@AndroidEntryPoint
class IntroSliderActivity : AppCompatActivity() {
    private lateinit var binding: ActivityIntroSliderBinding
    private lateinit var layouts: IntArray

    @Inject
    lateinit var preferencesService: PreferencesService

    override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)
        binding = ActivityIntroSliderBinding.inflate(layoutInflater)
        ViewCompat.setOnApplyWindowInsetsListener(binding.bottomBar) { v, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.updateLayoutParams<MarginLayoutParams> {
                bottomMargin = insets.bottom
            }
            WindowInsetsCompat.CONSUMED
        }
        setContentView(binding.root)
        layouts = intArrayOf(R.layout.intro_slider1, R.layout.intro_slider2)
        addBottomDots(0)
        val viewPagerAdapter = ViewPagerAdapter()
        binding.viewPager.adapter = viewPagerAdapter
        binding.viewPager.addOnPageChangeListener(viewListener)
        binding.btnSliderSkip.setOnClickListener {
            preferencesService.firstAppStart = true
            openMainActivity()
        }
        binding.btnSliderNext.setOnClickListener {
            val current = nextItem
            if (current < layouts.size) {
                binding.viewPager.currentItem = current
            } else {
                openMainActivity()
            }
        }

        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                openMainActivity()
            }
        })
    }

    private fun openMainActivity() {
        startActivity(Intent(this@IntroSliderActivity, MainActivity::class.java))
        finish()
    }

    private fun addBottomDots(position: Int) {
        val dots = arrayOfNulls<TextView>(layouts.size)
        val colorActive = resources.getIntArray(R.array.dot_active)
        val colorInactive = resources.getIntArray(R.array.dot_inactive)
        binding.layoutDots.removeAllViews()
        dots.indices.forEach { i ->
            val textView = TextView(this)
            textView.text = Html.fromHtml("&#8226;", Html.FROM_HTML_MODE_LEGACY)
            textView.textSize = 35f
            textView.setTextColor(colorInactive[position])
            dots[i] = textView
            binding.layoutDots.addView(textView)
        }
        if (dots.isNotEmpty()) {
            dots[position]!!.setTextColor(colorActive[position])
        }
    }

    private val nextItem: Int
        get() = binding.viewPager.currentItem + 1
    private val viewListener: OnPageChangeListener = object : OnPageChangeListener {
        override fun onPageScrolled(
            position: Int,
            positionOffset: Float,
            positionOffsetPixels: Int
        ) {
        }

        override fun onPageSelected(position: Int) {
            addBottomDots(position)
            if (position == layouts.size - 1) {
                binding.btnSliderNext.setText(R.string.proceed)
                binding.btnSliderSkip.visibility = View.INVISIBLE
                preferencesService.firstAppStart = true
            } else {
                binding.btnSliderNext.setText(R.string.next)
                binding.btnSliderSkip.visibility = View.VISIBLE
            }
        }

        override fun onPageScrollStateChanged(state: Int) {}
    }

    inner class ViewPagerAdapter : PagerAdapter() {
        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val layoutInflater = getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
            val view = layoutInflater.inflate(layouts[position], container, false)
            container.addView(view)
            return view
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            val view = `object` as View
            container.removeView(view)
        }

        override fun getCount(): Int {
            return layouts.size
        }

        override fun isViewFromObject(view: View, `object`: Any): Boolean {
            return view === `object`
        }
    }
}